lib/component.js

```diff
@@ -805,6 +819,109 @@ class Component extends WebComponent {
+
+    // vuejs like this.data.$data to reference "real" value of params
+    Object.defineProperty(this.params, `$params`, {
+      enumerable: false,
+      value: {},
+    });
+
+    for (const param of Object.keys(paramTypes)) {
+      // convert type shorthand to object
+      let paramType = paramTypes[param];
+      if (PARAM_TYPE_DEFAULTS.has(paramType)) {
+        paramType = {type: paramType};
+      }
+
+      // Ensure param type is valid
+      const type = paramType.type;
+      if (!PARAM_TYPE_DEFAULTS.has(type)) {
+        const typeString = typeof type === `function` ? type.name : String(type);
+        throw new Error(
+          `Invalid type: ${typeString} for attr: ${param} in paramType. ` +
+            `Only (${Array.from(PARAM_TYPE_DEFAULTS.keys())
+              .map((v) => `'${v.name}'`)
+              .join(` | `)}) is valid.`,
+        );
+      }
+
+      if (paramType.default && paramType.required) {
+        throw new Error(`${this}: param '${param}' cannot have both required and default`);
+      }
+
+      if ([Object, Array, Map, Set].includes(type) && !paramType.shouldUpdate) {
+        throw new Error(
+          `Object like params '${param}' on ${this.constructor.name} must have 'shouldUpdate' function passed in`,
+        );
+      }
+
+      const paramSchemaObj = {
+        type,
+        default: paramType.default ? paramType.default : PARAM_TYPE_DEFAULTS.get(type),
+        required: paramType.required ? paramType.required : false,
+        // by default use === for comparision
+        shouldUpdate: paramType.shouldUpdate || ((oldVal, newVal) => oldVal !== newVal),
+      };
+
+      // set default value for the params
+      this.params.$params[param] = paramSchemaObj.default;
+
+      this._paramTypes[param] = paramSchemaObj;
+    }
+
+    const paramDescription = Object.keys(paramTypes)
+      .map((param) => [
+        param,
+        {
+          get() {
+            return this.$params[param];
+          },
+        },
+      ])
+      .reduce((acc, cur) => {
+        acc[cur[0]] = cur[1];
+        return acc;
+      }, {});
+    Object.defineProperties(this.params, paramDescription);
+    Object.freeze(this.params);
+    return this._paramTypes;
+  }
+
+  _setParams(params) {
+    const toUpdate = {};
+    const updateOptions = {
+      cascade: false,
+      store: `params`,
+    };
+    for (const param in params) {
+      if (!this._paramTypes[param]) {
+        throw new Error(`${param} on ${this.constructor.name} is not defined in schema`);
+      }
+      const newParam = params[param];
+      const oldParam = this.params[param];
+      if (this._paramTypes[param].shouldUpdate(oldParam, newParam)) {
+        if (this._paramTypes[param].required && (newParam === undefined || newParam === null)) {
+          throw new Error(`${param} is required, ${newParam} passed in`);
+        }
+        toUpdate[param] = newParam;
+      }
+    }
+    if (this.initialized) {
```

Comment:
@themacguffinman this is what i've seen. so if no `enqueueMicroTask`, the `DOMPathcer` sometimes is not ready yet when _setParams is executed.

Reply:
That seems like the right behavior. It's what `this.update()` does. If you run `this.update()` before the component is initialized, it just skips the hook.

https://github.com/mixpanel/panel/blob/74dc4c21b3e1729ae5c38298d0fac64c23e58cc2/lib/component.js#L812-L827